// aspida/dist/cjs/getConfigs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfigs = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const defaultConfig = {
    input: (_a = ['api', 'apis'].find(input => fs_1.default.existsSync(input))) !== null && _a !== void 0 ? _a : 'api',
    baseURL: '',
    trailingSlash: false,
    outputEachDir: false,
    outputMode: 'all',
};
const getConfigs = (config = 'aspida.config.js') => {
    const configs = typeof config !== 'string'
        ? config
        : fs_1.default.existsSync(config)
            ? require(path_1.default.join(process.cwd(), config))
            : defaultConfig;
    return Array.isArray(configs)
        ? configs.map(c => ({ ...defaultConfig, ...c }))
        : [{ ...defaultConfig, ...configs }];
};
exports.getConfigs = getConfigs;
